File: bidmachine/src/main/java/io/bidmachine/examples/base/BaseExampleActivity.java
package io.bidmachine.examples.base;

import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.Switch;
import android.widget.TextView;

import io.bidmachine.BidMachine;

public class BaseExampleActivity extends AppCompatActivity {

    private FrameLayout adTypeContainer;
    private ProgressBar loadingProgressBar;
    private TextView currentAdStatus;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_base);

        adTypeContainer = findViewById(R.id.adTypecontainer);
        loadingProgressBar = findViewById(R.id.loadingProgressBar);
        currentAdStatus = findViewById(R.id.currentAdStatus);

        ((Switch) findViewById(R.id.testModeSwitcher)).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                BidMachine.setTestMode(isChecked);
            }
        });

        BidMachine.setTestMode(true);
    }

    @Override
    public void setContentView(int layoutResID) {
        if (adTypeContainer != null) {
            View view = LayoutInflater
                    .from(this)
                    .inflate(layoutResID, adTypeContainer, false);
            adTypeContainer.addView(view);
        } else {
            super.setContentView(layoutResID);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menu_item_kotlin_example) {
            try {
                PackageInfo packageInfo = getPackageManager().getPackageInfo(getPackageName(),
                        PackageManager.GET_ACTIVITIES | PackageManager.GET_META_DATA);
                for (ActivityInfo activityInfo : packageInfo.activities) {
                    if (activityInfo.metaData != null && activityInfo.metaData.getBoolean("bm-example-kotlin")) {
                        Intent intent = new Intent();
                        intent.setClassName(this, activityInfo.name);
                        startActivity(intent);
                        return true;
                    }
                }
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
            return false;
        } else if (item.getItemId() == R.id.menu_item_java_example) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    protected void toast(@NonNull String message) {
        Utils.toast(this, message);
    }

    protected void setDebugState(Status status) {
        setDebugState(status, null);
    }

    protected void setDebugState(Status status, String message) {
        if (loadingProgressBar != null) {
            loadingProgressBar.setVisibility(status.isLoading() ? View.VISIBLE : View.INVISIBLE);
        }
        if (currentAdStatus != null) {
            currentAdStatus.setText(status.getStatus());
        }
        if (TextUtils.isEmpty(message)) {
            toast("Banner Ads loaded");
        }
    }

    protected enum Status {
        Loading(true, "Loading"),
        Loaded(false, "Loaded"),
        LoadFail(false, "Load Fail"),
        Closed(false, "Closed"),
        Rewarded(false, "Rewarded");

        private boolean loading;
        private String status;

        Status(boolean loading, String status) {
            this.loading = loading;
            this.status = status;
        }

        public boolean isLoading() {
            return loading;
        }

        public String getStatus() {
            return status;
        }
    }

}

File: bidmachine/src/main/java/io/bidmachine/examples/base/BaseExampleActivity.kt
package io.bidmachine.examples.base

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Insets
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.WindowInsets
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.viewbinding.ViewBinding
import io.bidmachine.BidMachine
import io.bidmachine.examples.base.databinding.ActivityBaseBinding

abstract class BaseExampleActivity<Binding : ViewBinding> : AppCompatActivity() {

    private lateinit var baseBinding: ActivityBaseBinding

    protected lateinit var binding: Binding

    protected abstract fun inflate(inflater: LayoutInflater): Binding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState == null) {
            BidMachine.setTestMode(true)
        }

        binding = inflate(layoutInflater)
        baseBinding = ActivityBaseBinding.inflate(layoutInflater).apply {
            setContentView(root)
            setSupportActionBar(toolbar)
            toolbar.subtitle = "BidMachine Version - ${BidMachine.VERSION}"
            sTestMode.setOnCheckedChangeListener { _, isChecked ->
                BidMachine.setTestMode(isChecked)
            }
            adTypeContainer.addView(binding.root)
        }
        applyInsets(this, false)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_item_kotlin_example -> {
                try {
                    val activityInfos = packageManager.getPackageInfo(
                        packageName,
                        PackageManager.GET_ACTIVITIES or PackageManager.GET_META_DATA
                    )
                        .activities ?: return false
                    for (activityInfo in activityInfos) {
                        if (activityInfo.metaData != null && activityInfo.metaData.getBoolean(BM_KOTLIN)) {
                            val intent = Intent()
                            intent.setClassName(this, activityInfo.name)
                            startActivity(intent)
                            return true
                        }
                    }
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }
                return false
            }

            R.id.menu_item_java_example -> {
                finish()
                return true
            }

            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
    }

    protected fun toast(message: String?) {
        Utils.toast(this, message)
    }

    protected fun setDebugState(status: Status) {
        setDebugState(status, null)
    }

    protected fun setDebugState(status: Status, message: String?) {
        baseBinding.pbLoading.visibility = if (status.isLoading) {
            View.VISIBLE
        } else {
            View.INVISIBLE
        }
        baseBinding.tvStatus.text = status.status

        toast(message)
    }

    private fun applyInsets(activity: Activity, isRequestApplyInsets: Boolean) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.VANILLA_ICE_CREAM) {
            val contentView = baseBinding.root.apply {
                setPadding(0, 0, 0, 0)
            }
            val view = activity.window?.decorView ?: contentView
            setInsetsChanger(view, isRequestApplyInsets) {
                baseBinding.toolbar.setPadding(it.left, it.top, it.right, 0)
                view.setPadding(it.left, 0, it.right, it.bottom)
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private fun setInsetsChanger(view: View, isRequestApplyInsets: Boolean, action: (insets: Insets) -> Unit) {
        setInsetsChanger(
            view,
            isRequestApplyInsets,
            WindowInsets.Type.systemBars() or WindowInsets.Type.displayCutout(),
            action
        )
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    private fun setInsetsChanger(
        view: View,
        isRequestApplyInsets: Boolean,
        typeInsets: Int,
        action: (insets: Insets) -> Unit
    ) {
        view.setOnApplyWindowInsetsListener { _, windowInsets ->
            action(windowInsets.getInsets(typeInsets))
            windowInsets
        }
        if (isRequestApplyInsets) {
            view.requestApplyInsets()
        }
    }


    companion object {

        const val SOURCE_ID = "122"

        private const val BM_KOTLIN = "bm-example-kotlin"

    }

}